### DISCUSSION TITLE
JavascriptParser Hooks: Replace Parameter of function, async

### DISCUSSION THREAD

**Discussion by LukasBombach** [webpack/webpack#14956] [Q&A] [answered]
I am trying to asynchronously replace a parameter of a function call using the [`JavascriptParser Hooks` API](https://webpack.js.org/api/parser/).

I am importing a function called `createInlineScript` from my package `create-script/react` an I want to replace the parameter passed to it (a string) with another string that I can only get asynchronously once I know the orignal parameter

```tsx
import { createInlineScript } from "create-script/react";

const InlineScript = createInlineScript("../inline.ts");
//                                      ^^^^^^^^^^^^^^
//                                      I'm trying to replace this parameter with
//                                      another string
```

 So in this example I want to replace the path `"../inline.ts"` with the compiled sources of the file that path points to (using a child compiler).

I can replace the string, but only sync, using `parser.hooks.call`, so I cannot invoke a child compiler here, because compile code is `async`

```js
const pluginName = "create-script-plugin";
const IMPORT_TAG = Symbol("CreateScriptPlugin/tag");

class CreateScriptPlugin {

apply(compiler) {

    // Hooking into the parser
    compiler.hooks.compilation.tap(pluginName, compilation => {
      compiler.hooks.normalModuleFactory.tap(pluginName, factory => {
        factory.hooks.parser.for("javascript/auto").tap(pluginName, parser => {

          // I am tagging the imports of my lib so I can later hook into their usages
          parser.hooks.importSpecifier.tap(pluginName, (statement, source, id, name) => {
            if (source === 'create-script/react') {
              const ids = id === null ? [] : [id];
              parser.tagVariable(name, IMPORT_TAG, {
                name,
                source,
                ids,
                sourceOrder: parser.state.lastHarmonyImportOrder,
                await: statement.await,
              });
              return true;
            }
          });

          // Hooking into calls of the imported function
          parser.hooks.call.for(IMPORT_TAG).tap(pluginName, expression => {

            const arg = expression.arguments[0];

            // Replacing the argument of the function call with some random string
            // What I really want to do is create a `childCompiler` with
            // `compiler.createChildCompiler` and call `runAsChild` using the path
            // of the parameter I got here and use the result
            const dep = new ConstDependency(JSON.stringify("YADDA YADDA"), arg.range);
            dep.loc = arg.loc;
            parser.state.module.addPresentationalDependency(dep);

            return true;
          });
        });
      });
    });
  }
}

module.exports = CreateScriptPlugin;
```

**Comment by alexander-akait** [comment, marked as answer]
Sorry, it was not design for this. Why do not use babel/swc?

**Reply by LukasBombach** [reply]
Not using Babel because Next is phasing it out

SWC is certainly an option, I was only hesitant because I wanted to avoid require anymore tooling than needed because I'm afraid that if any plugin just adds their stuff (like their own version of swc) things just blow out of proportion.

It seems though that using swc seems the only reasonable option.

**Reply by LukasBombach** [reply]
One question before I give up though:

I am adding a `ConstDependency` inplace where I the function argument currently is and that Depency just generates a string everything has to be sync.

Could I maybe add another type of dependendy that would return the compiled code later?

Like use a another type of Dependency in `parser.hooks.call`

```js
parser.hooks.call.for(IMPORT_TAG).tap(pluginName, expression => {

  const pathArgument = expression.arguments[0];

  // use a dependency that takes a path and inserts the compiled source as string
  const dep = new CompileToStringDependency(pathArgument, arg.range);
  dep.loc = arg.loc;
  parser.state.module.addPresentationalDependency(dep);

  return true;
}
```

That dependendy would then compile the path and insert the compiled source as string

```js
class CompileToStringDependency {}

// compiles the path and inserts it
class CompileToStringDependencyTemplate {}
```

I wonder if that is possible or if such a dependency type already exists event

**Reply by alexander-akait** [reply]
Yes, it is possible, but you need to write template logic. Another solution - https://webpack.js.org/plugins/normal-module-replacement-plugin/, we even have example how to return different export, also we support `data:` URI, it can be done using them too. Why you need async here? Can you prepare all before passing it in webpack?

**Reply by LukasBombach** [reply]
How would that preparing work?

I do need to know the argument of the function call, which is a path to any file, once I know it, I can prepare it, I guess


`the_file_i_want_to_transform.tsx`
```tsx
import { createInlineScript } from "create-script/react";

const InlineScript = createInlineScript("../inline.ts");
//                                      ^^^^^^^^^^^^^^
//                                      I'm trying to replace this parameter with
//                                      another the compiled contents of that string

```

**Reply by alexander-akait** [reply]
I see... Can you provide code of `createInlineScript`? Is it just function returns another function?

**Reply by LukasBombach** [reply]
Basically yes. It is a function that return a react component that just inlines some JavaScript. 

The Idea is that developers can use `createInlineScript(path)` with a path as string argument and it will generate a react component that inlines the compiled source in a `<script>` tag by replacing the path argument with the compiled source without them "knowning" (not having to to do anything).


```tsx
import type { VFC } from "react";

export function createInlineScript<P>(source: string): VFC<P> {
  const InlineScript: VFC<P> = props => {
    const propsStr = JSON.stringify(props);
    const inlineJs = `(${source})(${propsStr})`;
    return <script dangerouslySetInnerHTML={{ __html: inlineJs }} />;
  };

  return InlineScript;
}
```

**Reply by LukasBombach** [reply]
The `props` thing here is so that you can provide data to the script when you use it

```tsx
const AdvertisementInit = createInlineScript("./adsInit.ts");

export const MyPage = () => (
  <head>
    <AdvertisementInit vendorId={12345} />
  </head>
);
```

during transpilation I am expecting a function as default export which takes the props as argument so I can pass the data to it in  line `6` where I do ``const inlineJs = `(${source})(${propsStr})`;``

**Reply by alexander-akait** [reply]
Why do not move logic inside `createInlineScript` and avoid this, i.e. `createInlineScript({ type: "advertisement", id: 1000 })`, this problem is real out of scope webpack, it is more about architecture, than bundling, you need bundling inside bundling based on arguments, honestly I think it might be worth considering other approaches and avoid unnecessary tooling

**Reply by LukasBombach** [reply]
I see and I get that. I don't need the props really, they just sort of "came for free" in that idea. But I wonder how to determerine from the app code which files should be transpiled to a string. Like with

```tsx
const InlineScript = createInlineScript("./file.ts");
```

I know it's the argument of `createInlineScript` which I need to transpile.

I would not know how to tell webpack to transpile the incoming file paths of `createInlineScript`.

I tried writing a loader, like

```tsx
import transpiledCodeAsString from "./file.ts";
```

but then I am not sure how to tell webpack to use that specific import to transform that to a string. Using `import x from "myloader!./file.ts"` breaks standards and intellisense in code editors.

**Reply by LukasBombach** [reply]
@alexander-akait thanks for your help. I think I understand the matter and the options I have now. It seems it to be possible—somehow— to use the Webpack parser to replace things, but at the same time, this is really, really effortful.

I now decided to go the cheapest way possible: I am regexing the incoming sourcecode in a loader and I will just find and wrap the path parameter with a `require()` funciton to that source. It is prone to erros when developers rename the imported function or assign it to some other variable, but I am willing to deal with those erros when they come up.

For now, that solution just saves time and complexity and will probably suffice for most use-cases
